---
title: "Microbiome vs Baseline Food Intake"
output: 
    flexdashboard::flex_dashboard:
        navbar:
            - { title: "About Fast Food Study", href: "../../website/_site/mcb.html", align: left }
        source_code: embed
        css: ../../css/styles.css
runtime: shiny 
---
<style>
.datatables{
    overflow: auto;
}
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, warning =F, error = F, message=F)
```

```{r, packages}
pkgs = c('plyr', 'dplyr','stringr','reshape2','tibble', 'plotly', 'DT',
         'ggplot2','ggsci')
for(pkg in pkgs){
    library(pkg, quietly=TRUE, verbose=FALSE, warn.conflicts=FALSE, 
            character.only=TRUE)
}
```

```{r}
load("../Rdata/blf_precalc.Rdata")
```

Inputs {.sidebar}
-------------------------------------

```{r}
inputPanel(
    selectInput('normal', 'Select A Transform Method: ',
                choices = names(blf_mcb_corr), 
                selected = "proportion"),
    selectInput('level','Select A Taxonomy Level: ',
                choices = names(blf_mcb_corr$reads), 
                selected = "Genus"),
    selectInput('food', 'Select A Food Category: ',
                choices = rownames(blf_data), 
                selected= rownames(blf_data)[1])
)
```

Column
-------------------------------------

### Chart 1

```{r}
renderPlotly({
    data = data.frame(
        FF = as.numeric(mcb_data_tx[[input$normal]][[input$level]][["FF"]][feature_selector(),]),
        Med = as.numeric(mcb_data_tx[[input$normal]][[input$level]][["Med"]][feature_selector(),]),
        Mix = as.numeric(mcb_data_tx[[input$normal]][[input$level]][["Mix"]][feature_selector(),]),
        food = as.numeric(blf_data[input$food,]),
        Subj = colnames(blf_data)
    ) %>%
        melt(id.var = c("Subj","food"),
             variable.name = "Group",
             value.name = "Value")
    p = ggplot(data, aes(x = food, y = Value)) +
        geom_point(aes(color = Subj), size=rel(2)) +
        facet_grid(.~Group) +
        scale_color_npg() +
        stat_smooth(method='lm', formula = y~x) +
        labs(
            x=input$food,
            y=feature_selector()
        )+
        theme_bw() +
        theme(
            legend.position = "none"
        )
    ggplotly(p, tooltip="all")
})
```

### Stats Table

```{r}
# renderDataTable({
#     data.frame(
#         pearson.pvalue = diet_corr[[input$transform]][[input$nutrient]][[input$taxo_level]]$pearson.pvalue,
#         Max = rowMaxs(as.matrix(edata_list[[input$transform]][[input$taxo_level]])),
#         Mean = rowMeans(edata_list[[input$transform]][[input$taxo_level]]),
#         Median = rowMedians(as.matrix(edata_list[[input$transform]][[input$taxo_level]])),
#         Zeros = apply(edata_list[[input$transform]][[input$taxo_level]],1, 
#                       function(x) sum(x==0))
#     ) %>%
#         datatable %>%
#         formatRound(columns = c("pearson.pvalue","Mean"), digits=3)
# })
```

```{r}
stats_table = reactive({
    alist = lapply(blf_mcb_corr[[input$normal]][[input$level]],
           function(group){
               group[[input$food]][,c("pearson.r","pearson.pvalue")]
           })
    data.frame(
        FF.r = alist$FF$pearson.r,
        FF.p = alist$FF$pearson.pvalue,
        Med.r = alist$Med$pearson.r,
        Med.p = alist$Med$pearson.pvalue,
        Mix.r = alist$Mix$pearson.r,
        Mix.p = alist$Mix$pearson.pvalue,
        Feature = rownames(alist$FF),
        stringsAsFactors = F
    )%>%
        arrange(Mix.p) %>%
        sapply(function(col){
            if(is.numeric(col))
                return(round(col, digits = 3))
            return(col)
        }) %>% 
        as.data.frame %>%
        column_to_rownames("Feature")
})
output$stats_table = renderDataTable(
    stats_table(),
    selection = list(mode = "single", selected = 1),
    server=T
)
feature_selector = reactive({
    rownames(stats_table())[input$stats_table_rows_selected]
})
dataTableOutput("stats_table")
```