---
title: "Microbiome Boxplot"
output: 
    flexdashboard::flex_dashboard:
        navbar:
            - { title: "About Fast Food Study", href: "../../website/_site/mcb.html", align: left }
        source_code: embed
runtime: shiny 
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, warning =F, error = F, message=F)
```

```{r, packages}
pkgs = c('dplyr', 'plyr','stringr','reshape2','tibble','plotly', "DT",
         'ggplot2','ggsci',"data.table")
for(pkg in pkgs){
    library(pkg, quietly=TRUE, verbose=FALSE, warn.conflicts=FALSE, 
            character.only=TRUE)
}
```

Inputs {.sidebar data-width=150}
----------------------------



```{r, load data and functions}
load('../Rdata/mcb_precalc.Rdata')
```


```{r input panel}
inputPanel(
    # the taxonomy level to view
    selectInput('level', 'Select a taxonomy level: ',
                choices = c(names(fdata), 'Feature'), selected = 'Genus'),
    # what kind of transformation to apply
    selectInput('transform', 'Select a transform method: ',
                choices = names(edata_list),
                selected = 'none')
)
```

Column 
-------------------------------------
    
### Boxplot

```{r}
renderPlotly({
    df = data.frame(
        Abundance = as.numeric(edata_list[[input$transform]][[input$level]][feature_selected(),]),
        Subj = pdata$StudyID,
        TX = pdata$Treatment,
        Day = pdata$Timepoint
    )
    p = ggplot(df, aes(x=Day, y=Abundance)) +
        geom_boxplot() +
        geom_point(aes(colour = Subj)) +
        geom_line(aes(group = Subj, colour = Subj)) +
        facet_grid(~TX) +
        theme_bw() +
        scale_color_npg()+
        theme(
            axis.title.x = element_blank(),
            axis.text.x = element_text(size=rel(1.5), color = 'black'),
            axis.ticks.x = element_blank(),
            axis.ticks.y = element_blank(),
            legend.position = 'none',
            strip.text = element_text(size=15)
            
        )
    ggplotly(p)
})
```

Column
-------------------------------------

###

<style>
    #stats_table{
        overflow: auto;
    }
</style>

```{r}
stats_table = reactive({
    stats_method = names(Differential_Abundance)[which(names(edata_list) == input$transform)]
    table = Differential_Abundance[[stats_method]][[input$level]] %>%
        as.data.frame() %>%
        rownames_to_column("Feature")
    if(input$transform == "reads"){
        table = arrange(table, pvalue)
    }else{
        table = arrange(table, P.Value)
    }
    table = sapply(table, function(col){
        if(!is.numeric(col)) return(col)
        round(col, 3)
    }) %>% as.data.frame%>%
        column_to_rownames("Feature")
})
output$stats_table = renderDataTable(
    stats_table(), 
    selection = list(mode = "single", selected = 1),
    server=T
    )
feature_selected = reactive({
    rownames(stats_table())[input$stats_table_rows_selected]
})
dataTableOutput("stats_table")
```
